Internal: REST API - V2 with a new convention [ED-627]

iNewLegend opened the pull request.

iNewLegend committed:
Data/Base: New - REST API Version 2.

iNewLegend committed:
Data/Base: New - REST API Version 2. (N1)

iNewLegend commented:
@matipojo what is the current status?

iNewLegend committed:
Data/Base: Remove - Sub-Controller.

iNewLegend committed:
Data/V2/Base: New - All-Children support sub-controllers.

iNewLegend committed:
Docs/Data/V2/Base: New - Controller documentation.

Nevoss commented on data/v2/base/base-route.php:
```diff
@@ -0,0 +1,340 @@
+<?php
```
register args for the default items endpoint

Nevoss commented on data/v2/base/endpoint/index.php:
```diff
@@ -0,0 +1,50 @@
+		return $this->controller->delete_items( $request );
+	}
+
+	public function register_item_route( $methods = WP_REST_Server::READABLE, $args = [], $route = '/' ) {
+		// Save 'id_arg_name' for 'sub controller, index endpoint'.
+		if ( ! empty( $args['id_arg_name'] ) && 'id' === $this->id_arg_name ) {
+			$this->id_arg_name = $args['id_arg_name'];
+		}
+
+		parent::register_item_route( $methods, $args, $route );
+	}
```
ensure method is required

iNewLegend replied:
Handled.

iNewLegend committed:
Merge remote-tracking branch 'Elementor/release/3.3.0' into feature/data/v2

iNewLegend committed:
Data/Base: New - REST API Version 2. (R1)

iNewLegend commented on data/manager.php:
```diff
@@ -94,13 +111,11 @@ public function register_controller( $controller_class_name ) {
-	public function register_controller_instance( $controller_instance ) {
-		// TODO: Validate instance.
-
+	public function register_controller_instance( /* Controller - TODO Remove - Support V2 */ $controller_instance ) {
```
Required for BC.

iNewLegend commented on data/v2/base/endpoint.php:
```diff
@@ -0,0 +1,158 @@
+	 */
+	private function get_ancestors() {
+		$ancestors = [];
+		$parent = $this;
```
Handled.

iNewLegend committed:
Data/Base: New - REST API Version 2. (R1N1)

iNewLegend committed:
Data/Base: New - REST API Version 2. (R1N2)

iNewLegend committed:
Data/Base: New - REST API Version 2. (R1N3)

iNewLegend committed:
Data/Base: New - REST API Version 2. (R1N4)

iNewLegend committed:
Docs/Data: Fix - controller.advance.md

matipojo commented on docs/data/base/v2/controller.md:
````diff
@@ -0,0 +1,326 @@
+    ```
+    * _Legend_
+       * *Controllers*: `House`, `Rooms`, `Doors`, `Keys`.
+       * *Sub-Controllers*: `Rooms`, `Doors`, `Keys`.
````
Controllers with a parent

iNewLegend replied:
Done.

matipojo commented on docs/data/base/v2/controller.md:
```diff
@@ -0,0 +1,326 @@
+    * When you need deeper endpoint that are parent of another endpoint.
+     
+* Why Controller exists?
+    * To border the best practice: 
```
```suggestion
    * To use the best practice as a resource.
```

iNewLegend replied:
Done.

matipojo commented on docs/data/base/v2/controller.md:
```diff
@@ -0,0 +1,326 @@
+        * `resource/{id}`
+        * `resource/{id}/resource/{id}`.
+        * `resource/endpoint/{id}`.
+    * To know how to serve you ( as resource ).
```
```suggestion
```

iNewLegend replied:
Done.

matipojo commented on docs/data/base/v2/controller.md:
```diff
@@ -0,0 +1,326 @@
+   *  To serve the `Controller` itself.
+
+* Why Sub-Endpoints exists?
+   *  Make it possible to simulate it as command, and attach a processor on it.
```
Add a notice about not to use it

iNewLegend replied:
Handled.

matipojo commented on docs/data/base/v2/controller.md:
```diff
@@ -0,0 +1,326 @@
+    * Endpoint represented as _route_
+   
+* Controller Simple advantages
+    * No requirement for `get_format` existence to support `class Processor` mechanism.
```
```suggestion
    * No requirement for `get_format`
    * To support `class Processor` mechanism.
```

iNewLegend replied:
Done.

matipojo commented on docs/data/base/v2/controller.md:
```diff
@@ -0,0 +1,326 @@
+        public function get_permission_callback( $request ) {
+           // Just for the example to work without extra permissions.
+           // In real word, it would check the user permission to the current endpoint.
+           return true;
```
```suggestion
           return false;
```

iNewLegend replied:
Done.

matipojo commented on docs/data/base/v2/controller.md:
```diff
@@ -0,0 +1,326 @@
+        public function get_permission_callback( $request ) {
+                // Just for the example to work without extra permissions.
+                // In real word, it would check the user permission to the current resource.
+                return true;
```
```suggestion
                return false;
```

iNewLegend replied:
Done.

matipojo commented on docs/data/base/v2/controller.md:
```diff
@@ -0,0 +1,326 @@
+        public function get_permission_callback( $request ) {
+            // Just for the example to work without extra permissions.
+            // In real word, it would check the user permission to the current resource.
+            return true;
```
False

iNewLegend replied:
Done.

matipojo commented on tests/phpunit/elementor/data/test-manager.php:
```diff
@@ -1,31 +1,18 @@
-	public function tearDown() {
-		parent::tearDown();
-
 		$this->manager->kill_server();
```
Try to move it to the base

iNewLegend replied:
Cannot it should support backwards compatibility

matipojo commented on tests/phpunit/elementor/data/v2/base/mock/template/traits/mock-bypass-permission.php:
```diff
@@ -0,0 +1,23 @@
+<?php
+namespace Elementor\Tests\Phpunit\Elementor\Data\V2\Base\Mock\Template\Traits;
+
+trait Mock_Bypass_Permission {
```
```suggestion
trait Mock_Force_Permission {
```

iNewLegend replied:
Done.

matipojo commented on tests/phpunit/elementor/data/v2/base/mock/template/traits/mock-bypass-permission.php:
```diff
@@ -0,0 +1,23 @@
+
+trait Mock_Bypass_Permission {
+	public $bypass_permission_value = true;
+	public $is_permission_bypassed = false;
```
```suggestion
	public $is_permission_forced = false;
```

iNewLegend replied:
Done.

matipojo commented on tests/phpunit/elementor/data/v2/base/mock/template/traits/mock-bypass-register.php:
```diff
@@ -0,0 +1,24 @@
+<?php
+namespace Elementor\Tests\Phpunit\Elementor\Data\V2\Base\Mock\Template\Traits;
+
+trait Mock_Bypass_Register {
```
Explain

iNewLegend replied:
Done.

matipojo commented on tests/phpunit/elementor/data/v2/base/mock/template/traits/mock-items.php:
```diff
@@ -0,0 +1,86 @@
+trait Mock_Items {
+	use Mock_Test_Data;
+
+	public function get_items( $request ) {
```
Try to wrap with a reflection class.

iNewLegend replied:
In which method or where I can warp it in Trait?

matipojo commented on tests/phpunit/elementor/data/v2/base/mock/template/traits/mock-name-type.php:
```diff
@@ -0,0 +1,16 @@
+<?php
+namespace Elementor\Tests\Phpunit\Elementor\Data\V2\Base\Mock\Template\Traits;
+
+trait Mock_Name_Type {
```
```suggestion
trait Mock_Random_Name_Type {
```

iNewLegend replied:
Done.

matipojo commented on tests/phpunit/elementor/data/v2/base/test-endpoint.php:
```diff
@@ -0,0 +1,254 @@
+		$descendant_endpoint->set_test_data( 'get_items', 'valid' );
+
+		// Act.
+		$data = $this->manager->run( $descendant_endpoint->get_full_command() );
```
Assert also the full command

iNewLegend replied:
Done.

matipojo commented on tests/phpunit/elementor/data/v2/base/test-sub-controller.php:
```diff
@@ -0,0 +1,125 @@
+		$this->assertEquals( $controller->get_name() . '/' . $sub_controller->get_name(), $actual );
+	}
+
+	public function test_execute_simple() {
```
```suggestion
	public function test_run_endpoint__simple() {
```
Fix all text names.

iNewLegend replied:
Done.

matipojo reviewed (changes requested):
Create also a PR for Submission that uses v2.

iNewLegend commented on data/v2/base/base-route.php:
```diff
@@ -0,0 +1,340 @@
+	/**
+	 * Constructor.
+	 *
+	 * run `$this->>register()`.
```
```suggestion
	 * run `$this->register()`.
```

iNewLegend committed:
Data/Base: New - REST API Version 2. (R2).

iNewLegend commented on data/v2/base/processor.php:
```diff
@@ -0,0 +1,32 @@
+	exit; // Exit if accessed directly
+}
+
+abstract class Processor {
```
Handled.

iNewLegend commented on data/manager.php:
```diff
@@ -83,7 +100,7 @@ private function set_cache( $key, $value ) {
-	 * @return \Elementor\Data\Base\Controller
+	 * @return \Elementor\Data\V2\Base\Controller
 	 */
 	public function register_controller( $controller_class_name ) {
```
Handled.

iNewLegend commented on data/manager.php:
```diff
@@ -94,13 +111,11 @@ public function register_controller( $controller_class_name ) {
-	public function register_controller_instance( $controller_instance ) {
-		// TODO: Validate instance.
-
+	public function register_controller_instance( /* Controller - TODO Remove - Backwards compatibility for V1 */ $controller_instance ) {
```
Handled used `register_controller` instead.

iNewLegend commented on data/manager.php:
```diff
@@ -324,10 +339,10 @@ public function run_processors( $processors, $filter_by_class, $data ) {
 	 * @return \WP_REST_Response
 	 */
-	private function run_request( $endpoint, $args, $method ) {
+	public function run_request( $endpoint, $args = [], $method = \WP_REST_Server::READABLE ) {
```
Done

iNewLegend commented on data/v2/base/controller.php:
```diff
@@ -0,0 +1,402 @@
+
+		$parent_name = $this->get_parent_name();
+		if ( $parent_name ) {
+			$this->act_as_children( $parent_name );
```
Explain in method signature

iNewLegend commented on data/v2/base/controller.php:
```diff
@@ -0,0 +1,402 @@
+	/**
+	 * Get parent controller name.
+	 *
+	 * @note: If `get_parent_name()` provided, controller will work as children controller.
```
Done.

iNewLegend commented on data/v2/base/controller.php:
```diff
@@ -0,0 +1,402 @@
+	protected function register_processor( $processor_class ) {
+		$processor_instance = new $processor_class( $this );
+
+		// TODO: Validate processor instance.
```
Done

iNewLegend commented on data/v2/base/controller.php:
```diff
@@ -0,0 +1,402 @@
+		$this->register_processors();
+	}
+
+	private function act_as_children( $parent_name ) {
```
Done.

iNewLegend commented on data/v2/base/controller.php:
```diff
@@ -0,0 +1,402 @@
+			trigger_error( "Cannot find parent controller: '$parent_controller_name'", E_USER_ERROR ); // phpcs:ignore
+		}
+
+		$this->parent->children [] = $this;
```
Done.

iNewLegend commented on data/v2/base/endpoint.php:
```diff
@@ -0,0 +1,158 @@
+				$ancestors [] = $current;
+			}
+
+			if ( ! is_callable( [ $current, 'get_parent' ] ) ) {
```
Done.

iNewLegend commented on docs/data/base/v2/controller.advance.md:
````diff
@@ -0,0 +1,176 @@
+```
+
+## The example available to download:
+ [Click here to download plugin example](https://drive.google.com/file/d/1pYijAc7SJdBPeFJoy0wwJHEmtO-t_Ja9/view?usp=sharing)
````
Done.

iNewLegend commented on docs/data/base/v2/controller.md:
```diff
@@ -0,0 +1,326 @@
+       
+* What are Sub-Controllers & Sub-Endpoints?
+    * Terms used to describe the flow.
+    * Sub-Controller are *Controller* which is children of another controller.
```
Done.

iNewLegend committed:
Merge branch 'release/3.4.0' into feature/data/v2

iNewLegend committed:
Data/Globals/Controller: Fix - Use new V2.

matipojo reviewed (changes requested):
Please add:
- Response Convention
- Error Handling

iNewLegend committed:
JS API/CommandData: New - ApplyAfterCatch.

There is two ways of handling failed fetch(s).
1. from the command it self using `applyAfterCatch()`.
2. Using catch hook ( $e.hooks ).

1. is used when no ui effects needed.
2. used when it have to effects the UI.

iNewLegend committed:
Data/V2: New - Error handling.

iNewLegend committed:
Data/V2: New - Error handling. (N1).

iNewLegend commented on core/common/assets/js/api/modules/command-data.js:
```diff
@@ -202,16 +202,19 @@ export default class CommandData extends CommandBase {
 		// TODO: If the errors that returns from the server is consistent remove the '?' from 'e'
-		const status = e?.data?.status || 0;
+		// eslint-disable-next-line camelcase
+		const httpErrorCode = e?.data?.status || 0;
```
```suggestion
		const httpErrorCode = e?.data?.status || 501;
```

iNewLegend commented on core/common/assets/js/api/modules/command-data.js:
```diff
@@ -202,16 +202,19 @@ export default class CommandData extends CommandBase {
 	onCatchApply( e ) {
 		// TODO: If the errors that returns from the server is consistent remove the '?' from 'e'
-		const status = e?.data?.status || 0;
+		// eslint-disable-next-line camelcase
```
```suggestion
```

iNewLegend committed:
Data/V2: New - Error handling. (N2).

Nevoss commented on core/common/assets/js/api/modules/command-data.js:
```diff
@@ -202,16 +202,19 @@ export default class CommandData extends CommandBase {
 	onCatchApply( e ) {
 		// TODO: If the errors that returns from the server is consistent remove the '?' from 'e'
-		const status = e?.data?.status || 0;
+		const httpErrorCode = e?.data?.status || 501;
```
Why 501 and not 500?

iNewLegend replied:
The default exception converted to error 500.

In case you add more generic error(s) (by extending `Elementor\Data\V2\Base\Exceptions\Data_Exception`)
and you forget to give the right HTTP code, the error will be '501 not implemented'.

matipojo reviewed (changes requested):
Please pull from 3.5 and fix CI errors.

iNewLegend committed:
Merge remote-tracking branch 'Elementor/release/3.5.0' into feature/data/v2

iNewLegend committed:
TemplateLibrary: Change - Move to REST v2.

iNewLegend committed:
Fix: PHPLint.

iNewLegend committed:
Fix: PHPLint.

matipojo reviewed (changes requested):
Fix globals/screenshotter

matipojo commented on data/manager.php:
```diff
@@ -56,6 +57,13 @@ public static function instance() {
+		add_action( 'elementor/init', function () {
+			Plugin::$instance->modules_manager->get_modules( 'dev-tools' )->deprecation->deprecated_function(
+				'Elementor\Data\Manager',
+				'3.4.0',
```
```suggestion
				'3.5.0',
```

iNewLegend replied:
Handled https://github.com/elementor/elementor/pull/14205/commits/c5a31de9983c6d4014dce754b47d3c0d4fe8cf11

matipojo commented on data/manager.php:
```diff
@@ -56,6 +57,13 @@ public static function instance() {
 	}
 
 	public function __construct() {
+		add_action( 'elementor/init', function () {
```
The notice shouldn't appear in core (only for 3rd that still uses it)

iNewLegend replied:
Handled https://github.com/elementor/elementor/pull/14205/commits/c5a31de9983c6d4014dce754b47d3c0d4fe8cf11

matipojo commented on data/v2/base/controller.php:
```diff
@@ -0,0 +1,401 @@
+	 *
+	 */
+	public function __construct() {
+		$this->namespace = Manager::ROOT_NAMESPACE . '/v' . Manager::VERSION;
+		$this->rest_base = Manager::REST_BASE . $this->get_name();
```
Make it extandable

iNewLegend replied:
Handled https://github.com/elementor/elementor/pull/14205/commits/0bd0b5556a56f1f98cb1b4dffbfba5568d22d57d

matipojo commented on data/v2/base/exceptions/error-500.php:
```diff
@@ -0,0 +1,17 @@
+<?php
+namespace Elementor\Data\V2\Base\Exceptions;
+
```
Add for all files

if ( ! defined( 'ABSPATH' ) ) {
	exit; // Exit if accessed directly
}

iNewLegend replied:
Handled https://github.com/elementor/elementor/pull/14205/commits/168bd99c71bf7a66472b48dd573266fee062a103

matipojo commented on docs/data/base/v2/controller.advance.md:
````diff
@@ -0,0 +1,176 @@
+```
+
+## The full example:
+ [https://github.com/iNewLegend/elementor-rest-api-example](https://github.com/iNewLegend/elementor-rest-api-example)
````
Move to here

iNewLegend replied:
Handled https://github.com/elementor/elementor/pull/14205/commits/ae28e06a4c823ac50eb4b05c55ed10ae3c36569f

matipojo commented on includes/plugin.php:
```diff
@@ -504,16 +505,15 @@ class Plugin {
-	 * @var \Core\Data\Manager
+	 * @var \Elementor\Data\Manager
 	 */
 	public $data_manager;
```
Add `@deprecated` + create an instance only on `__call`

iNewLegend replied:
Handled https://github.com/elementor/elementor/pull/14205/commits/c5a31de9983c6d4014dce754b47d3c0d4fe8cf11

matipojo replied:
Please create an instance only on __call

matipojo commented on tests/phpunit/elementor/data/test-manager.php:
```diff
@@ -270,11 +270,7 @@ public function test_run() {
-			'template-library/templates' => 'template-library/templates',
-		], $this->manager->command_formats );
+		// Deprecated.
+		$this->assertEquals( [], $this->manager->command_formats );
```
Delete

iNewLegend replied:
Done https://github.com/elementor/elementor/pull/14205/commits/1fc9c234fec02fcdebbe76fad78d86edb9f136d1

rodolphebertozzo commented on core/editor/data/globals/endpoints/base.php:
```diff
@@ -1,35 +1,38 @@
-				'global_not_found',
-				__( 'The Global value you are trying to use is not available.', 'elementor' ),
-				[ 'status' => Exceptions::NOT_FOUND ]
+			throw new Error_404( __( 'The Global value you are trying to use is not available.', 'elementor' ),
```
esc

rodolphebertozzo commented on core/editor/data/globals/endpoints/base.php:
```diff
@@ -40,7 +43,7 @@ public function create_item( $id, $request ) {
 
 		if ( ! isset( $item['title'] ) ) {
-			return new \WP_Error( 'invalid_title', 'Invalid title' );
+			return new Data_Exception( __( 'Invalid title', 'elementor' ), 'invalid_title' );
```
esc

iNewLegend committed:
Data\Controller: Change - Make namepsace & version extendable.

iNewLegend committed:
Data/V2: Fix - ABSPATH.

iNewLegend committed:
Merge remote-tracking branch 'Elementor/release/3.5.0' into feature/data/v2

iNewLegend committed:
Data/V1: Remove - Tests.

iNewLegend committed:
Revert "Data/V1: Remove - Tests."

This reverts commit 1fc9c234fec02fcdebbe76fad78d86edb9f136d1.

iNewLegend committed:
Globals: Change - Move to V2.

iNewLegend commented:
> Fix globals/screenshotter

Handled https://github.com/elementor/elementor/pull/14205/commits/7971d06d92e49ea74a22591be69d48e6b2b7f3c5

iNewLegend commented on data/manager.php:
```diff
@@ -4,6 +4,7 @@
 
```
Move all core data to use V2.

iNewLegend replied:
Handled https://github.com/elementor/elementor/pull/14205/commits/f60bf5710ddfb9ae44b0e95271c9427ee566d01a

iNewLegend replied:
https://github.com/elementor/elementor/pull/14205/commits/6aba6f78bd29408f5e1f7bb44e25dbaedb1f3c2a

iNewLegend replied:
https://github.com/elementor/elementor/pull/14205/commits/0e3a49b332c3af4711be0fda721a98eab85b6113

iNewLegend committed:
Data/V1: Fix - Deprecation.

iNewLegend committed:
Data/V2: New - Move full example, to current repo.

iNewLegend committed:
Modules/KitLibrary: Change - Move to Data V2.

iNewLegend committed:
Modules/KitLibrary: Fix - Move to Data V2.

iNewLegend committed:
TemplateLibrary: Fix - Register controller via data v2.

iNewLegend committed:
Plugin: Add - @Deprecated annotation for v1.

iNewLegend committed:
Merge branch 'release/3.5.0' into feature/data/v2

matipojo commented on core/app/modules/kit-library/data/taxonomies/controller.php:
```diff
@@ -2,45 +2,37 @@
 		];
 	}
+
+	public function register_endpoints() {
```
Make it default (not abstract)

iNewLegend replied:
Handled: https://github.com/elementor/elementor/pull/14205/commits/86a144e3ede79b7e613a4aafcfc4b0a22b7dace8

matipojo commented on data/manager.php:
```diff
@@ -4,6 +4,7 @@
 
 use Elementor\Core\Base\Module as BaseModule;
 use Elementor\Data\Base\Processor;
+use Elementor\Plugin;
```
Check if needed

iNewLegend replied:
Handled https://github.com/elementor/elementor/pull/14205/commits/c7c5bdeb3e68fb9473e9a2a8382b521c1d88d606

matipojo commented on tests/phpunit/elementor/data/base/mock/processor/controller.php:
```diff
@@ -1,11 +0,0 @@
-<?php
```
Please restore the tests until the pro will be moved to the v2

iNewLegend replied:
Handled https://github.com/elementor/elementor/pull/14205/commits/401180279f6ddbcb71fc3ca70c327dd33a821438

iNewLegend committed:
Data/Manager: Remove - Unrequited import.

iNewLegend committed:
Data/V2: Remove - Abstract from `register_endpoints()`.

iNewLegend commented:
@matipojo `Please create an instance only on __call`, cannot do it devTools module, no available at the time.

iNewLegend committed:
Remove `$data_manager` v1 from 'plugin.php'.

iNewLegend committed:
Fix: 'controller.md'
